service/internal/asg/asg.go

```diff
@@ -0,0 +1,248 @@
+package asg
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/aws-operator/pkg/awstags"
+	"github.com/giantswarm/aws-operator/service/controller/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/key"
+	"github.com/giantswarm/aws-operator/service/internal/asg/internal/cache"
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+
+	Stack        string
+	TagKey       string
+	TagValueFunc func(cr key.LabelsGetter) string
+}
+
+type ASG struct {
+	k8sClient k8sclient.Interface
+
+	instancesCache *cache.Instances
+
+	stack        string
+	tagKey       string
+	tagValueFunc func(cr key.LabelsGetter) string
+}
+
+func New(config Config) (*ASG, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+
+	if config.Stack == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Stack must not be empty", config)
+	}
+	if config.TagKey == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.TagKey must not be empty", config)
+	}
+	if config.TagValueFunc == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.TagValueFunc must not be empty", config)
+	}
+
+	a := &ASG{
+		k8sClient: config.K8sClient,
+
+		instancesCache: cache.NewInstances(),
+
+		stack:        config.Stack,
+		tagKey:       config.TagKey,
+		tagValueFunc: config.TagValueFunc,
+	}
+
+	return a, nil
+}
+
+func (a *ASG) Any(ctx context.Context, obj interface{}) (string, error) {
+	cr, err := meta.Accessor(obj)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	instances, err := a.cachedInstances(ctx, cr)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return asgNameFromInstance(instances[0]), nil
+}
+
+func (a *ASG) Drainable(ctx context.Context, obj interface{}) (string, error) {
+	cr, err := meta.Accessor(obj)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	instances, err := a.cachedInstances(ctx, cr)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	var names []string
+	{
+		m := map[string]struct{}{}
+
+		for _, i := range instances {
+			m[asgNameFromInstance(i)] = struct{}{}
+		}
+
+		for k := range m {
+			names = append(names, k)
+		}
+	}
+
+	asgs, err := a.cachedASGs(ctx, names)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return asgs[0].Name, nil
+}
+
+func (a *ASG) cachedInstances(ctx context.Context, cr metav1.Object) ([]*ec2.Instance, error) {
```

Comment:
is it a good idea to cache instances? they are pretty volatile so we might need to have up-to-date info every time it runs

Reply:
Caching means to remember the results within one reconciliation loop. On the next loop we look them up again. This should be fine.

Reply:
ack